### PR TITLE
debug/profiling: initial API

### PR DESCRIPTION
**Pull request by microbuilder**
**WIP, do not merge.**

Provided for discussion purposes only during the next API call.

## Design Goal

This PR is an attempt to provide an instrumentation framework to allow reasonably accurate measurement of execution time for arbitrary blocks of code. 

The initial back-end for this will be the [CoreSight `DWT`](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0314h/Chdiicbc.html) peripheral block, which is standard on Cortex M3/M4/M7 and higher cores.

The goal, however, is to have a hardware-agnostic framework that can fall back to less reliable system timers or architecture/device-specific options where appropriate.

## Why is this needed?

There is currently a [Tracing API](https://docs.zephyrproject.org/latest/guides/tracing/index.html) in Zephyr, but this is targetted more at tools like Segger SystemView and while it is very useful for tracing tasks in the system, it isn't necessarily appropriate to instrument arbitrary chunks of code when trying to optimise mathematical functions, etc.

Making use of dedicated peripheral blocks like CoreSight and DWT provide a far more accurate view of real execution time, with less processing and tool overhead, and no proprietary tool dependencies.

Instrumenting a section of code with `DWT` is as simple as resetting the cycle counter, executing the code in question, then reading the cycle counter to know how many clock cycles it took (with a margin of error in the tens of clock cycles for processing overhead):

```
// Macro to initialise, reset and enable the cycle counter.
// This can be used for rough timing and performance tests
// by resetting the cycle counter before a function, and
// then reading the value after with "int count = DWT->CYCCNT"
//
//    DWT_RESET_CYCLECOUNTER;
//    ... do something
//    int count = DWT->CYCCNT;

#define DWT_RESET_CYCLECOUNTER    do { CoreDebug->DEMCR = CoreDebug->DEMCR | 0x01000000;  \
                                       DWT->CYCCNT = 0;                                   \
                                       DWT->CTRL = DWT->CTRL | 1 ; } while(0)
```

### Timing source for platform-agnostic backend

How should timing backends for instrumentation be handled on non M3/M4/M7/M23/M33 systems that don't have an equivalent to DWT? 

Presumably we can fall back to a high accuracy timer as a worst-case but universal option, even if this has a higher margin of error, but provides a global option plus `DWT` as an architecture-specific example.

### Integrate `stats` into the Instrumentation API?

Zephyr already has a [`stats`](https://github.com/zephyrproject-rtos/zephyr/tree/master/subsys/stats) module that provides high-quality instrumentation data for field-debugging, tracking the number of times user-defined events occur on the system (sensor reads, bus errors, etc.). While this is different than the aim of the instrumentation API, which is accurately measuring clock cycles for code optimisation efforts, `stats` does fall in the instrumentation domain.

Should stats and clock cycle counts be merged into a single `instrument` API?

### CODE REVIEW

**Comment by microbuilder** [comment]
Support should be added for the following [DWT registers](https://developer.arm.com/docs/ddi0337/e/system-debug/dwt/summary-and-description-of-the-dwt-registers) as well, in addition of course to abstracting away the actual timer backends to specific architectures:

- [`DWT_EXCCNT`: Exception Overhead Count Register](https://developer.arm.com/docs/ddi0337/e/system-debug/dwt/summary-and-description-of-the-dwt-registers#BABIEJJJ)
- [`DWT_SLEEPCNT`: Sleep count register](https://developer.arm.com/docs/ddi0337/e/system-debug/dwt/summary-and-description-of-the-dwt-registers#BABGIDEE)

**Comment by bfrog** [comment]
This is very useful to me in the short term, as I was looking to start measuring the performance of various zio parts. I think it would be very convenient if there was a instrument macro which allowed you to instrument a block, something like 

```instrument_block(some_block_name, { do_my_complex_thing(); });```

which seems completely feasible, I'd prefer to use a static string for the instrument tasks myself rather than an integer id, doesn't seem like it would change much and would help when there's a lot of logs going on.

further out, it would be amazing to have something that lets you generate flame graph usable data of instrumented code sections to see where all your time is going

**Comment by microbuilder** [comment]
Potentially some overlap with https://github.com/zephyrproject-rtos/zephyr/issues/15676 if sleep cycle count is implemented.

**Review by dleach02 — COMMENTED** [review]
You have an implied usage here where one could have multiple "task" structures to keep track of multiple timed events but the problem is that the functions reset the baseline DWT counters which will result in multiple counters being messed up. If your intent is to allow multiple counters to be tracked then you should probably have a starting value in your counter and keep a differential count. Problem though is now you could run into a wrap event. 

An alternative is to not expose multiple "task" structures and keep an internal task structure so you only have a single counter to track internally. Personally, I think it might be more useful to allow multiple task structures.

Another observation is that the baseline functions might be more useful to return a cycle count value (and a function to return sleep, etc... if you like the comment from @microbuilder) and allow the user to decide what to do with the cycle count. I had previously used this DWT to embed cycle count in every log message.

File: subsys/debug/profiling/profiling.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#include "profiling_priv.h"
+
+LOG_MODULE_REGISTER(profiling, CONFIG_PROFILING_LOG_LEVEL);
+
+int dbg_prof_init(void)
+{
+	return 0;
+}
+
+int dbg_prof_task_init(struct dbg_prof_task *task)
```
[the remark this answers is not in the document]
Comment (microbuilder) [reply subsys/debug/profiling/profiling.c]: I agree with you here ... the initial idea was to have something architecture nuetral falling backing to one or more generic timers when DWT isn't present (M3+), but just to test out the proof of concept I ended up using DWT directly which, as you mentionned, it restricted to a single instance without performing all kind of dubious acrobatics.

It probably makes more sense to limit this to a single timer/counter instance, and hide the details in the functions.

[the remark this answers is not in the document]
Comment (microbuilder) [reply subsys/debug/profiling/profiling.c]: Multiple timers or multiple tasks with the same timer handling roll-over would be nice, but in most instances you likely only want to instrument a single chunk of code at a time anyway, but the `update` function could be used to display interim values and some manual calculations could be done for intermediate measurements?

I also need to migrate this to either inline functions in a header of macros, and compensate for the overhead in the 'update' functions to stop and continue the timer before and after the update call.

**Comment by microbuilder** [comment]
Closing as stale, and should be rethought for cross platform support. Will reopen as bandwidth permits.

**Pull request closed without merging**